src/main/java/org/folio/orders/events/handlers/ReceiptStatusConsistency.java

```diff
@@ -0,0 +1,195 @@
+            future.completeExceptionally(e);
+            return null;
+          });
+      } else {
+        logger.debug("Receipt status of pieces and po line are consistent");
+        future.complete(null);
+      }
+    })
+      .exceptionally(e -> {
+        logger.error("The error happened getting a piece by id {}", pieceIdUpdate, e);
+        future.completeExceptionally(e);
+        return null;
+      });
+
+    // Now wait for all operations to be completed and send reply
+    completeAllFutures(ctx, futures, message);
+  }
+
+
+  private CompletableFuture<PoLine.ReceiptStatus> calculatePoLineReceiptStatus(PoLine poLine,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+
+    if (pieces.isEmpty())
+      return completedFuture(poLine.getReceiptStatus());
+    else {
+      return getPiecesQuantityByPoLineAndStatus(ReceivingStatus.EXPECTED, pieces)
+        .thenCompose(expectedQty -> calculatePoLineReceiptStatus(expectedQty, poLine, pieces))
+        .exceptionally(e -> {
+          logger.error("The expected receipt status for PO Line '{}' cannot be calculated", e, poLine.getId());
+          return null;
+        });
+    }
+  }
+
+  private CompletableFuture<ReceiptStatus> calculatePoLineReceiptStatus(int expectedPiecesQuantity, PoLine poLine,
+      List<org.folio.rest.acq.model.Piece> pieces) {
+    if (!isCheckin(poLine) && expectedPiecesQuantity == 0) {
+      return CompletableFuture.completedFuture(FULLY_RECEIVED);
+    }
+    // Partially Received: In case there is at least one successfully received
+    // piece
+    if (StreamEx.of(pieces)
+      .anyMatch(piece -> ReceivingStatus.RECEIVED == piece.getReceivingStatus())) {
+      return CompletableFuture.completedFuture(PARTIALLY_RECEIVED);
+    }
+    // Pieces were rolled-back to Expected. In this case we have to check if
+    // there is any Received piece in the storage
+    return getPiecesQuantityByPoLineAndStatus(ReceivingStatus.RECEIVED, pieces)
+      .thenApply(receivedQty -> receivedQty == 0 ? AWAITING_RECEIPT : PARTIALLY_RECEIVED);
+  }
+
+  private CompletableFuture<Integer> getPiecesQuantityByPoLineAndStatus(ReceivingStatus receivingStatus, List<Piece> pieces) {
+    Integer count = 0;
+    for (Piece piece : pieces) {
+      if (piece.getReceivingStatus()
+        .value()
+        .equalsIgnoreCase(receivingStatus.toString())) {
+        count++;
+      }
+    }
+    final Integer expectedQty = count;
+    return CompletableFuture.supplyAsync(() -> expectedQty);
+  }
+
+  CompletableFuture<PieceCollection> getPieces(String query, String path, HttpClientInterface httpClient, Context ctx,
+      Map<String, String> okapiHeaders, Logger logger) {
+    CompletableFuture<PieceCollection> future = new VertxCompletableFuture<>(ctx);
+    try {
+      String queryParam = isEmpty(query) ? EMPTY : "&query=" + encodeQuery(query, logger);
+      String endpoint = String.format(path, LIMIT, OFFSET, queryParam, LANG);
```

Comment:
I think this means you'll only get up to the hardcoded `LIMIT` number of records.  TBH it's probably not going to be a common occurrence for there to be more than 500 Pieces for a given poLine, but I think it's possible and should be handled appropriately.

Reply:
To avoid hardcoding, the only way I know is to get total piece records is to query `/orders-storage/pieces` which is another call to the orders-storage module. Is there a better way to get a total piece count without hitting orders-storage module?

Reply:
Instead of asking for 500, ask for a smaller, more reasonable number, say 10 - 20.  In the response you'll get the total number of results.  Make additional calls while increasing the offset by the appropriate amount until you have all of the results.

Reply:
I have updated logic to get all pieces by using Integer.MAX_VALUE as the limit. This should take care even if we have more than 500 pieces. Offset/pageable will introduce unnecessary complexity by making recurrent calls to GET all pieces if `getTotalRecords` exceeds a smaller reasonable set LIMITS.

Reply:
@folio-org/acquisitions does anyone know off-hand if we use paging in other places in acquisitions?  Using a giant limit param doesn't seem ideal to me, if there are enough records and the storage module is slow enough you run the risk of the request timing out.  It also places additional memory strain on the storage module as it would need to load all of those records into memory before returning them.

Reply:
As far as I know we do not use this pattern yet. Other option is to use 0 as limit parameter which basically means return no records at all but only count of such records i.e. `totalRecords`. For example, [CheckinReceivePiecesHelper#getPiecesQuantityByPoLineAndStatus(poLineId, receivingStatus)](https://github.com/folio-org/mod-orders/blob/master/src/main/java/org/folio/rest/impl/CheckinReceivePiecesHelper.java#L563) is designed such way.

Reply:
OK thanks @piotr-kalashuk.  Something to keep in mind going forward I guess